presto-hive/src/main/java/com/facebook/presto/hive/util/FilterPushdownUtils.java

```diff
@@ -0,0 +1,532 @@
+/*
```

Comment:
@yingsu00 I had thought about it as well, but I saw a few drawbacks.  HiveFilterPushdown has some class fields like HiveTransactionManager, HivePartitionManager, etc. which are not required in IcebergFilterPushdown since it has its own versions like IcebergTransactionManager. Also, in HiveFilterPushdown's pushdownFilter method, we have HiveMetadata whereas IcebergFilterPushdown we have implementations of IcebergAbstractMetadata. 

IMO directly extending HiveFilterPushdown class might not be ideal, if we want I think we can try creating a new base class extracting the common fields between Hive and Iceberg and then derive both HiveFilterPushdown and IcebergFilterPushdown from that base class. 

Let me know what you think

Reply:
@imjalpreet Making a base class is ok. I think most of the content can be in the base class and only very minimal difference between the two derived classes.

In filterPushdown(), metadata is only used at two places: metadata.getColumnHandles(session, tableHandle) and metadata.getTableLayout(...) and both are on ConnectorMetadata. So it's ok to change the metadata parameter from HiveMetadata and IcebergAbstractMetadata to ConnectorMetadata.

Now let's look at transactionManager and icebergTransactionManager. They are only used to get the Metadata in getMetadata(TableHandle tableHandle) method:
```
    protected IcebergAbstractMetadata getMetadata(TableHandle tableHandle)
    {
        ConnectorMetadata metadata = icebergTransactionManager.get(tableHandle.getTransaction());
        checkState(metadata instanceof IcebergAbstractMetadata, "metadata must be IcebergAbstractMetadata");
        return (IcebergAbstractMetadata) metadata;
    }
```
The downcast was only to make sure the metadata is IcebergAbstractMetadata, but icebergTransactionManager would always return IcebergXXXMetadata so returning IcebergAbstractMetadata instead of ConnectorMetadata is not very necessary. If we make it just return a ConnectorMetadata, then the getMetadata() method can have the same signature and be put to the base class.

partitionManager was also only used once. 
```
HivePartitionResult hivePartitionResult = partitionManager.getPartitions(metastore, tableHandle, constraint, session);
```
Similar to getMetadata(), we can have an abstract method in the base class, depending on how you express the Iceberg partition result. If that way doesn't work we can also break filterPushdown into multiple sections, with most sections having the same content.

So please go ahead and extract common base class.

Reply:
@yingsu00 I have refactored the filter pushdown and introduced a new abstract class.